src/Sparrow.Server/Threading/AsyncGuard.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using Sparrow.Binary;
+
+namespace Sparrow.Server.Threading;
+
+/// <summary>
+/// This class allows many "visitors" to enter/exit it, with minimal concurrency / contention between them
+/// It also allows a *single* thread to move guard to closed state, with the following rules:
+/// * Once the guard is marked closed, no future entries our allowed
+/// * Closing the guard may block until there are no more visitors
+/// </summary>
+public class AsyncGuard : IDisposable
+{
+    private readonly ulong[] _coreMarkers;
+    private readonly int _mask;
+    private readonly ManualResetEvent _event;
+    private bool _locked;
+    private const ulong WriterLockMask = 0x8000_0000_0000_0000;
+
+    public AsyncGuard()
+    {
+        int len = Bits.PowerOf2(Environment.ProcessorCount * 2);
+        _mask = len - 1;
+        _coreMarkers = new ulong[len];
+        _event = new ManualResetEvent(false);
+    }
+
+    public bool TryEnter(out int idx)
+    {
+        idx = Environment.CurrentManagedThreadId & _mask;
+        ref var cur = ref _coreMarkers[idx];
+        ulong copy = cur;
+        while (true)
+        {
+            if ((copy & WriterLockMask) != 0)
+                return false; // already taken
+
+            var result = Interlocked.CompareExchange(ref cur, copy + 1, copy);
+            if (result == copy)
+                return true; // got it successfully
+
+            copy = result;
+        }
+    }
+
+    public void Exit(int idx)
+    {
+        // we use the idx here to ensure that we can call this from a different thread
+        ref var cur = ref _coreMarkers[idx];
+        ulong copy = cur;
+        ulong masked;
+        while (true)
+        {
+            masked = copy & WriterLockMask;
+            ulong update = (copy & ~WriterLockMask) - 1;
+            if (update > copy)
```

Comment:
This check requires to check only magnitude, Copy may have  the mask in case this happens after the guard is locked.

Reply:
Note that we _exclude_ the `WriterLockMask`, so the top bit is cleared.
This is a check to ensure that if we have two calls to `Exit()` that shouldn't be, which will cause us to flip from `0` to `-1`

Reply:
That's correct. Because the first part does not requires cooperation from any party.